libs/bundles/libs-bundles-opensaml/pom.xml

```diff
@@ -1,223 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```

Comment:
After previous discussions about removing this, I ended up going ahead and doing it. The reason was that the Servicemix OpenSAML bundle now has no version constraint on the Guava version is uses. See: https://github.com/apache/servicemix-bundles/blob/master/opensaml-3.4.5/pom.xml#L63

This means it will pick up the Guava version that DDF uses automatically; we don't need to force the version here. Note, though, that we have to make sure that there is only one Guava version available or there will be uses constraint violations.

I'm open to keeping this bundle if we think that's a better option. But these changes work just fine without it.

Reply:
I vote we remove it.

Reply:
(❗ ) If only one Guava version can be available, then this may break some Camel features. Check out the Camel features file: 
```
<feature name="camel-protobuf" version="${camel.version}" start-level="50">
        <feature version="${camel.version}">camel-core</feature>
        <bundle dependency="true">mvn:com.google.protobuf/protobuf-java/${protobuf-version}</bundle>
        <bundle dependency="true">mvn:com.google.protobuf/protobuf-java-util/${protobuf-version}</bundle>
        <bundle dependency="true">mvn:commons-io/commons-io/${commons-io-version}</bundle>
        <bundle dependency="true">mvn:com.google.guava/guava/${camel.guava.version}</bundle>
        <bundle dependency="true">mvn:com.google.code.gson/gson/${gson.version}</bundle>
        <bundle>mvn:org.apache.camel/camel-protobuf/${camel.version}</bundle>
    </feature>
```
It appears the Camel version of Guava is lower than DDF's version of Guava. 
```
<camel.guava.version>19.0</camel.guava.version>
```

Reply:
Turns out this is actually a lot more nuanced than I was thinking. Interestingly, installing the feature you linked also causes uses constraint violations on 2.26.0. It's not just these changes - *DDF cannot have two Gauva versions installed at the same time*. Here's a sample of the constraint violations you see if it does:

```
2020-10-05T08:56:24,035 | DEBUG | tures-3-thread-1 | ResolverImpl                     | ternal.resolver.Slf4jResolverLog   45 | 17 - org.apache.karaf.features.core - 4.2.9 | Candidate permutation failed due to a conflict between imports; will try another if possible. (Uses constraint violation. Unable to resolve resource solr-factory-impl [solr-factory-impl/2.26.0] because it is exposed to package 'com.google.common.collect' from resources com.google.guava [com.google.guava/25.1.0.jre] and com.google.guava [com.google.guava/25.1.0.jre] via two dependency chains.

Chain 1:
  solr-factory-impl [solr-factory-impl/2.26.0]
    import: (&(osgi.wiring.package=com.google.common.collect)(version>=25.1.0)(!(version>=26.0.0)))
     |
    export: osgi.wiring.package: com.google.common.collect
  com.google.guava [com.google.guava/25.1.0.jre]

Chain 2:
  solr-factory-impl [solr-factory-impl/2.26.0]
    import: (&(osgi.wiring.package=org.codice.ddf.platform.util)(version>=2.26.0)(!(version>=3.0.0)))
     |
    export: osgi.wiring.package=org.codice.ddf.platform.util; uses:=com.google.common.io
  platform-util [platform-util/2.26.0]
    import: (&(osgi.wiring.package=com.google.common.io)(version>=25.1.0)(!(version>=26.0.0)))
     |
    export: osgi.wiring.package: com.google.common.io; uses:=com.google.common.collect
    export: osgi.wiring.package=com.google.common.collect
  com.google.guava [com.google.guava/25.1.0.jre])
```

A couple notes:
* This isn't actually an error, just a Karaf debug message. I found that interesting - there are non-fatal uses constraint violations. However, while it doesn't break anything Karaf takes significantly longer to resolve bundle wiring. On my machine, a `profile:install standard` took about 6 minutes compared to 1 minute normally.
* The error message here is a bit confusing since it references 25.1.0.jre for both dependency chains. It's really an issue with the uses constrains being satisfied by both versions of Guava present.
* Part of the problem is DDFs own leaky abstractions: platform-util exposes Guava in its API. Meaning anything that uses Guava and platform-util will have uses constraints violations if two Guava versions are present.

The only reason DDF 2.26.0 doesn't have these issues by default is that it doesn't install multiple Guava versions. That surprised me because the CXF and DDF versions differ. I dug into it a bit and it's because CXF uses the dependency flag in its feature descriptor. From the Karaf mailing list, here's a description of the dependency flag:

> The role of the dependency attribute is to mark that a bundle is a dependency. If a dependency is already satisfied (an existing bundle already exports the same packages/version) then it doesn't get installed. This behavior happens if the declared feature resolver is installed (e.g. obr is installed).

The wrapped OpenSAML bundle imports Guava packages with the version matching the DDF Guava version. So when when Karaf installs OpenSAML, it sees that the imports are already satisfied and so doesn't try to install the CXF Guava version (note, CXF itself doesn't import Guava. It only pulls it in to satisfy OpenSAML). Now that we're using the OpenSAML bundle without a version constraint, it appears Karaf doesn't think the imports are already satisfied and so installs the new version.

To summarize: with these changes, the only difference from master is that the CXF and DDF Guava versions must match to prevent uses constraint violations. With or without these changes, the presence of multiple Guava versions is problematic and will cause non-fatal uses constraint violations and significantly longer installation times.

Reply:
That's some quality background and investigation. Even though we can get the container to start I still believe that 6 minutes to start is just as bad as a borked container that doesn't work at all because you have to wait far too long (synchronously) to learn if you've succeeded or failed. So for future readers of this - know that wiring issues like the above are basically fatal even if the container is capable of recovering.

To add to the above ☝️ the OpenSAML bundle needed to be wrapped because, as of October of last year, no new versions existed without the transitive version constraint. As @SmithJosh pointed out, this has recently been resolved with several new artifacts released from servicemix. 

Additionally, between the time I upgraded Guava and this change, the class path complexity has increased. Platform-util is now a bundle which itself exposes Guava. 
Guava upgrade: https://github.com/codice/ddf/pull/5379
Platform util: https://github.com/codice/ddf/pull/5776

This means that, while the Camel feature may have worked fine last year, it no longer does. Likely because of a new class path collision with platform util's Guava and the code that Camel needs. 

Given the wrapper is no longer serving any purpose I'm +1 to remove it now, as well. 👍

Reply:
Thanks @Lambeaux. Completely agree with this:
> Even though we can get the container to start I still believe that 6 minutes to start is just as bad as a borked container that doesn't work at all because you have to wait far too long (synchronously) to learn if you've succeeded or failed

I created https://github.com/codice/ddf/issues/6365 to fix it